snuba/query/snql/joins.py

```diff
@@ -0,0 +1,188 @@
+from __future__ import annotations
+from typing import MutableMapping, NamedTuple, Optional, Sequence, Union
+
+from snuba.datasets.entities import EntityKey
+from snuba.query.data_source.join import (
+    IndividualNode,
+    JoinClause,
+    JoinCondition,
+    JoinConditionExpression,
+    JoinRelationship,
+)
+from snuba.query.data_source.simple import Entity as QueryEntity
+from snuba.query.parser.exceptions import ParsingException
+
+
+class RelationshipTuple(NamedTuple):
+    lhs: IndividualNode[QueryEntity]
+    relationship: str
+    rhs: IndividualNode[QueryEntity]
+    data: JoinRelationship
+
+
+class Node:
+    """
+    This class is a linked list of entities to join together. Each Node holds its own IndividualNode
+    that ultimately is added to a JoinClause. If the Node is not the root of the linked list, it
+    will also contain the join conditions that describe which LHS entity the Node should be joined
+    and the specifics of that join.
+
+    As new joins are added, they are inserted into the linked list in the correct order so the join
+    clause will be nested correctly.
+
+    """
+
+    def __init__(
+        self,
+        entity_data: IndividualNode[QueryEntity],
+        relationship: Optional[JoinRelationship] = None,
+    ) -> None:
+        self.entity_data = entity_data
+        self.relationship = relationship
+        self.child: Optional[Node] = None
+        self.join_conditions: Sequence[JoinCondition] = []
+
+    @property
+    def entity(self) -> EntityKey:
+        assert isinstance(self.entity_data.data_source, QueryEntity)
+        return self.entity_data.data_source.key
+
+    def push_child(self, node: Node) -> None:
+        # This happens here to ensure we use the correct alias in the columns.
+        self.build_join_conditions(node)
+        if not self.child:
+            self.child = node
+            return
+
+        old_child = self.child
+        self.child = node
+        if node.child is None:
+            node.child = old_child
+            return
+
+        # iterate down list and push old_child to end
```

Comment:
I think this comment is still valid:

Not sure why you need to put the nodes in this specific order.
It seems that this takes a list as input and it inserts that entire list after the current node (self) and before the tail of the list represented by self.
Was there any reason not to append the whole input to the end ?
If yes could add a comment explaining the reason. (I feel that explaining the "why" something is donne in a way is more useful that explaining "what is about to happen")

Reply:
It doesn't matter for correctness whether the new child is put before or after the old child. I think practically doing it this way would have a tiny (very tiny) performance benefit since in most cases the new child is shorter than the old child.

Reply:
Ok, up to you. If you keep it this way, I would document on the why, because that is the first thing that comes to mind when reading this people ask themselves why isn't it done in the most intuitive way.

I would not bother too much about that specific tiny performance gain considering the context this fits into (several passes and rewrites over the entire query, and the actual clickhouse query). 
This code is going to be executed only once per query over generally 2-3 nodes. It is not like the cost of rebuilding an expression during transformations that can happen multiple times per node in the ast per query. That case would be an interesting one to profile instead.